Add logging definition file for OS log entries that should be forwarded from the WebContent process to the UI process

pvollan opened the pull request:
#### a443055848fd199a509bfb778e40094d32822c53
<pre>
Add logging definition file for OS log entries that should be forwarded from the WebContent process to the UI process
<a href="https://rdar.apple.com/137601908">rdar://137601908</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=282115">https://bugs.webkit.org/show_bug.cgi?id=282115</a>

Reviewed by Chris Dumez.

Currently, we are forwarding preprocessed OS logs as strings from the WebContent process to the UI process in Lockdown Mode,
where they are sent to the logging daemon. Logging preprocessed strings is effectively disabling an important log size
optimization in the logging system. To preserve this optimization, all logging should use a literal format string, and
provide the data arguments when using the OS log API. This patch resolves this by adding logging definition files that are
used to generate code to send IPC messages from the WebContent process to the UI process, where the OS log API is called
with a literal format string, and the arguments provided over IPC. This patch adds one logging definition file for WebCore,
and one for WebKit, both with the name LogMessages.in.

This is the current format of the logging definition file:

&lt;Identifier&gt;, &lt;Format string&gt;, &lt;Parameters&gt;, &lt;Log type (DEFAULT, INFO, ERROR, FAULT)&gt;, &lt;Category (Default, Process, Loading, etc)&gt;

Here is an example:

DOCUMENTLOADER_STOPLOADING, &quot;[pageID=%llu, frameID=%llu, isMainFrame=%d] DocumentLoader::stopLoading&quot;, (uint64_t, uint64_t, int), DEFAULT, Network

To add logging in the WebContent process that will be forwarded to the UI process in the efficient way, you will need
to add an entry to one of the logging definition files, and then pass the identifier to one of the new logging macros
- RELEASE_LOG_FORWARDABLE, RELEASE_LOG_ERROR_FORWARDABLE, etc.

Based on the logging definition files, the file LogStream.messages.in is generated. This file is used as input to the
standard script that generates code to send messages over IPC. The log forwarding is using streaming IPC, which is
based on shared memory, and is very efficient.

Additionally, two new log client classes are created, WebCore::LogClient and WebKit::LogClient. The scripts added in this
patch will generate virtual functions in these classes for every log message in the definitions file. Using one of the
new logging macros (RELEASE_LOG_FORWARDABLE, etc.), will result in calling one of these virtual functions, which will
send the IPC to the UI process. WebKit::LogClient is inheriting from WebCore::LogClient, and is implementing the sending
of IPC for both WebCore and WebKit logs.

This patch enables log forwarding for all Cocoa platforms, not just in Lockdown Mode.

Page load benchmarks results are neutral with this change for the configurations tested so far, and the sample count
is also neutral in PLT, which indicates that there should be no power regression from this change. Power benchmark
results are neutral on both iOS and macOS.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::m_remote):
(WebCore::HTMLMediaElement::~HTMLMediaElement):
(WebCore::HTMLMediaElement::insertedIntoAncestor):
(WebCore::HTMLMediaElement::didFinishInsertingNode):
(WebCore::HTMLMediaElement::removedFromAncestor):
(WebCore::HTMLMediaElement::currentMediaTime const):
(WebCore::HTMLMediaElement::mediaPlayerTimeChanged):
(WebCore::HTMLMediaElement::mediaPlayerRateChanged):
(WebCore::HTMLMediaElement::mediaPlayerEngineUpdated):
(WebCore::HTMLMediaElement::updatePlayState):
(WebCore::HTMLMediaElement::visibilityStateChanged):
(WebCore::HTMLMediaElement::setShouldDelayLoadEvent):
(WebCore::HTMLMediaElement::setBufferingPolicy):
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::mediaPlayerRenderingModeChanged):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::stopLoading):
(WebCore::DocumentLoader::attachToFrame):
(WebCore::DocumentLoader::detachFromFrame):
(WebCore::DocumentLoader::startLoadingMainResource):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::changeLocation):
(WebCore::FrameLoader::loadURLIntoChildFrame):
(WebCore::FrameLoader::prepareForLoadStart):
(WebCore::FrameLoader::loadFrameRequest):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::loadWithNavigationAction):
(WebCore::FrameLoader::loadWithDocumentLoader):
(WebCore::FrameLoader::stopAllLoaders):
(WebCore::FrameLoader::setDocumentLoader):
(WebCore::FrameLoader::setPolicyDocumentLoader):
(WebCore::FrameLoader::setProvisionalDocumentLoader):
(WebCore::FrameLoader::setState):
(WebCore::FrameLoader::transitionToCommitted):
(WebCore::FrameLoader::checkLoadCompleteForThisFrame):
(WebCore::FrameLoader::continueLoadAfterNavigationPolicy):
* Source/WebCore/loader/ProgressTracker.cpp:
(WebCore::ProgressTracker::progressStarted):
(WebCore::ProgressTracker::progressCompleted):
(WebCore::ProgressTracker::finalProgressComplete):
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::willSendRequestInternal):
* Source/WebCore/loader/SubresourceLoader.cpp:
(WebCore::SubresourceLoader::willSendRequestInternal):
(WebCore::SubresourceLoader::didReceiveResponse):
(WebCore::SubresourceLoader::didFinishLoading):
(WebCore::SubresourceLoader::didFail):
(WebCore::SubresourceLoader::willCancel):
* Source/WebCore/platform/Logging.cpp:
(WebCore::logClient):
* Source/WebCore/platform/Logging.h:
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::addSession):
(WebCore::PlatformMediaSessionManager::removeSession):
(WebCore::PlatformMediaSessionManager::sessionWillEndPlayback):
(WebCore::PlatformMediaSessionManager::sessionCanProduceAudioChanged):
(WebCore::PlatformMediaSessionManager::maybeActivateAudioSession):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::updateSessionState):
(WebCore::MediaSessionManagerCocoa::clientCharacteristicsChanged):
(WebCore::MediaSessionManagerCocoa::sessionCanProduceAudioChanged):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::didReceiveInvalidMessage):
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Platform/Sources.txt:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/Scripts/generate-message-receiver.py:
(main):
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/API/c/mac/WKWebArchiveResource.cpp:
(WKWebArchiveResourceCopyURL):
(WKWebArchiveResourceCopyMIMEType):
* Source/WebKit/Shared/APIWebArchiveResource.h:
* Source/WebKit/Shared/APIWebArchiveResource.mm:
(API::WebArchiveResource::create):
(API::WebArchiveResource::WebArchiveResource):
(API::WebArchiveResource::url):
(API::WebArchiveResource::mimeType):
(API::WebArchiveResource::textEncoding):
(API::WebArchiveResource::URL): Deleted.
(API::WebArchiveResource::MIMEType): Deleted.
* Source/WebKit/Shared/LogStream.cpp:
(WebKit::LogStream::~LogStream):
(WebKit::LogStream::setup):
* Source/WebKit/Shared/LogStream.h:
(WebKit::LogStream::create):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::setupLogStream):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::finishLaunchingProcess):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::scheduleLoad):
(WebKit::WebLoaderStrategy::scheduleLoadFromNetworkProcess):
* Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
(WebKit::WebResourceLoader::WebResourceLoader):
(WebKit::WebResourceLoader::willSendRequest):
(WebKit::WebResourceLoader::didReceiveResponse):
(WebKit::WebResourceLoader::didReceiveData):
(WebKit::WebResourceLoader::didFinishResourceLoad):
(WebKit::WebResourceLoader::serviceWorkerDidNotHandle):
(WebKit::WebResourceLoader::didFailResourceLoad):
(WebKit::WebResourceLoader::didBlockAuthenticationChallenge):
(WebKit::WebResourceLoader::stopLoadingAfterXFrameOptionsOrContentSecurityPolicyDenied):
(WebKit::WebResourceLoader::didReceiveResource):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::navigationActionData const):
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/cocoa/LaunchServicesDatabaseManager.mm:
(WebKit::LaunchServicesDatabaseManager::handleEvent):
(WebKit::LaunchServicesDatabaseManager::waitForDatabaseUpdate):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformInitializeWebProcess):
(WebKit::WebProcess::setupLogStream):
(WebKit::WebProcess::sendLogOnStream):
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

Canonical link: <a href="https://commits.webkit.org/287665@main">https://commits.webkit.org/287665@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a947863352f340df9cda0d5f27d9f867a75a8c5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80372 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59378 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33888 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84893 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31354 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82483 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68440 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7681 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62818 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20626 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83441 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52896 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73182 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43122 "Passed tests") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/79829 "webkitpy-tests (failure)") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50213 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27337 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29813 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/73352 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71352 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27854 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86327 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/79431 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7597 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5374 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71097 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7772 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69020 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70337 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14337 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13281 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/101838 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12444 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7559 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13079 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24817 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-no-cjit (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7398 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10918 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9204 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Platform/WebKitLogClient.h:
```diff
@@ -0,0 +1,71 @@
+
+namespace WebKit {
+
+class WebKitLogClient : public WebCore::WebCoreLogClient {
```
`final`

pvollan replied:
Thanks for reviewing!

I will address these issues in the upcoming patch.

cdumez commented on Source/WebKit/Platform/WebKitLogClient.h:
```diff
@@ -0,0 +1,71 @@
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    WebKitLogClient(IPC::StreamClientConnection&, const LogStreamIdentifier&);
+    virtual ~WebKitLogClient() { };
```
No need for trailing `;`

cdumez commented on Source/WebKit/Platform/WebKitLogClient.h:
```diff
@@ -0,0 +1,71 @@
+    WebKitLogClient(IPC::StreamClientConnection&, const LogStreamIdentifier&);
+    virtual ~WebKitLogClient() { };
+
+    Ref<IPC::StreamClientConnection> logStreamConnection() const { return m_logStreamConnection; }
```
This either needs to return a `IPC::StreamClientConnection&` or be named `protectedLogStreamConnection()`

cdumez commented on Source/WebKit/Platform/WebKitLogClient.h:
```diff
@@ -0,0 +1,71 @@
+
+    Ref<IPC::StreamClientConnection> logStreamConnection() const { return m_logStreamConnection; }
+    LogStreamIdentifier logStreamIdentifier() const { return m_logStreamIdentifier; }
+    Lock& logStreamLock() { return m_logStreamLock; }
```
Seems like this should use `WTF_RETURNS_LOCK()`.

As a matter of fact, is there a way to make use of our macros in ThreadSafetyAnalysis.h to validate the thread safety model at compile time?

cdumez replied:
For example, should `logStreamConnection()` use `WTF_REQUIRES_LOCK()`

cdumez commented on Source/WebKit/Scripts/generate-derived-log-sources.py:
```diff
@@ -0,0 +1,143 @@
+            arguments_string = log_declarations_module.get_arguments_string(parameters, log_declarations_module.PARAMETER_LIST_INCLUDE_TYPE | log_declarations_module.PARAMETER_LIST_INCLUDE_NAME | log_declarations_module.PARAMETER_LIST_MODIFY_CSTRING)
+            file.write("    virtual void " + function_name + "(" + arguments_string + ")\n")
+            file.write("    {\n")
+            file.write("        m_logStreamLock.lock();\n")
```
Please use `Locker`

cdumez commented on Source/WebKit/Scripts/generate-derived-log-sources.py:
```diff
@@ -0,0 +1,143 @@
+            file.write("        if (RefPtr connection = logStreamConnection())\n")
+            parameters_string = log_declarations_module.get_arguments_string(parameters, log_declarations_module.PARAMETER_LIST_INCLUDE_NAME)
+            file.write("            connection->send(Messages::LogStream::" + function_name + "(" + parameters_string + "), m_logStreamIdentifier);\n")
+            file.write("        m_logStreamLock.unlock();\n")
```
Not needed if you use `Locker`.

cdumez commented on Source/WebKit/Platform/WebKitLogClient.h:
```diff
@@ -0,0 +1,71 @@
+private:
+    Ref<IPC::StreamClientConnection> m_logStreamConnection;
+    LogStreamIdentifier m_logStreamIdentifier;
+    static Lock m_logStreamLock;
```
I'm a little unclear why the Lock is `static` here. Is there ever more than one log client? Even if there were several, would we really need synchronization between all instances?

cdumez commented on Source/WebKit/Shared/LogStream.h:
```diff
@@ -40,16 +40,22 @@ namespace WebKit {
 public:
-    LogStream() = default;
-    ~LogStream();
+    static RefPtr<LogStream> create() { return adoptRef(new LogStream()); }
```
Should be:
```
static Ref<LogStream> create() { return adoptRef(*new LogStream); }
```

3 fixes.

cdumez commented on Source/WebKit/Shared/LogStream.h:
```diff
@@ -40,16 +40,22 @@ namespace WebKit {
 
 private:
+    LogStream() = default;
+    ~LogStream();
```
Why is the destructor private? I guess this means LogStreams are never destroyed? Which is weird because the LogStream destructor does do work.

cdumez replied:
Still private?

pvollan replied:
Ah, good catch! Fixing now.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
```diff
@@ -262,7 +262,9 @@
 void WebProcessProxy::setupLogStream(uint32_t pid, IPC::StreamServerConnectionHandle&& serverConnection, LogStreamIdentifier logStreamIdentifier, CompletionHandler<void(IPC::Semaphore& streamWakeUpSemaphore, IPC::Semaphore& streamClientWaitSemaphore)>&& completionHandler)
 {
-    m_logStream.setup(pid, WTFMove(serverConnection), logStreamIdentifier, WTFMove(completionHandler));
+    m_logStream = LogStream::create();
```
Better written as:
```
Ref logStream = LogStream::create();
logStream->setup(pid, WTFMove(serverConnection), logStreamIdentifier, WTFMove(completionHandler));
m_logStream = WTFMove(logStream);
```

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -862,24 +862,34 @@ static void prewarmLogs()
 #if PLATFORM(IOS_FAMILY)
         prewarmLogs();
 #endif
+        webkitLogClient() = WTF::makeUnique<WebKitLogClient>(*logStreamConnection.get(), logStreamIdentifier);
```
We shouldn't need `WTF::` or `.get()`

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -862,24 +862,34 @@ static void prewarmLogs()
         prewarmLogs();
 #endif
+        webkitLogClient() = WTF::makeUnique<WebKitLogClient>(*logStreamConnection.get(), logStreamIdentifier);
+        webCoreLogClient() = WTF::makeUnique<WebKitLogClient>(*logStreamConnection.get(), logStreamIdentifier);
```
ditto.

cdumez replied:
Maybe WebKitLogClient should be ref-counted and we could use the same instance for `webCoreLogClient()` and `webkitLogClient()`? Seems a bit wasteful to allocate 2 objects. Or we could keep using unique_ptr but have a single `logClient()` in WebCore, used by both WebKit and WebCore?

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -862,24 +862,34 @@ static void prewarmLogs()
-    if (RefPtr logStreamConnection = m_logStreamConnection)
-        logStreamConnection->send(Messages::LogStream::LogOnBehalfOfWebContent(logChannel, logCategory, logString, type), m_logStreamIdentifier);
+    if (RefPtr logStreamConnection = webkitLogClient()->logStreamConnection()) {
+        webkitLogClient()->logStreamLock().lock();
```
Should use `Locker` but also, What's the locking protecting us against here?

pvollan replied:
This function is called on the same thread as the OS log API is called on, so we need to synchronize the access to the log stream connection.

cdumez commented on Source/WebKit/Platform/Logging.cpp:
```diff
@@ -27,8 +27,22 @@
+#if ENABLE(LOGD_BLOCKING_IN_WEBCONTENT)
+namespace WebKit {
+
+RefPtr<WebKitLogClient>& webkitLogClient()
```
Why do we need `webkitLogClient()`, why can't we use `webcoreLogClient()` everywhere, including in WebKit?

pvollan replied:
That's a good point. The reason for this approach is that `WebCoreLogClient` only has the virtual functions for the WebCore logs, generated from `WebCoreLogMessages.in`. The class 'WebKitLogClient' has the virtual functions for the WebKit logs, generated from 'WebKitLogMessages.in', as well as the virtual functions from `WebCoreLogClient`, since it inherits from that class. I chose to have separate `.in` files in WebKit and WebCore, because I think it would be good to avoid that WebCore has information about WebKit logs.

cdumez commented on Source/WebKit/Platform/WebKitLogClient.h:
```diff
@@ -0,0 +1,77 @@
+class WebKitLogClient final : public WebCore::WebCoreLogClient {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    static RefPtr<WebKitLogClient> create(IPC::StreamClientConnection& connection, const LogStreamIdentifier& identifier)
```
Why is this still returning a RefPtr instead of a Ref?

pvollan replied:
Ah, good catch, fixing now!

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -862,24 +862,33 @@ static void prewarmLogs()
 #endif
+        RefPtr logClient = WebKitLogClient::create(*logStreamConnection, logStreamIdentifier);
+        webkitLogClient() = logClient;
+        webCoreLogClient() = logClient;
```
Still unclear why we need 2 clients (and ref-counting)

pvollan replied:
It is a very good point. I am looking into if we can have a single log client, and still have a separation between WebCore and WebKit log definitions.

cdumez replied:
One idea could be that the macros in WebKit/ could downcast the WebCoreClient into a WebKitClient before calling whatever log function.

cdumez replied:
And then I'd rename `webCoreLogClient()` into `logClient()`

pvollan replied:
That's a very good idea! I was looking into solving this by how we generate the declarations, but down casting seems a much more elegant approach :)

cdumez commented on Source/WebKit/Platform/Logging.h:
```diff
@@ -28,6 +28,50 @@
+#include "WebKitLogClient.h"
+
+#define RELEASE_LOG_FORWARDABLE(category, logMessage, ...) do { \
+    if (auto* client = static_cast<WebKitLogClient*>(WebCore::logClient().get())) \
```
We should be using `downcast` in new code, as suggested earlier.

pvollan replied:
Ah, yes, good point, will fix!

cdumez commented on Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
```diff
@@ -128,7 +128,7 @@ void StreamServerConnection::didClose(Connection&)
 }
 
-void StreamServerConnection::didReceiveInvalidMessage(Connection&, MessageName, int32_t)
+void StreamServerConnection::didReceiveInvalidMessage(Connection&, MessageName name, int32_t index)
```
Why this change?

pvollan replied:
Ah, good catch, I had some debug code here, and I forgot to remove the parameter names again :) Will fix!

cdumez commented on Source/WebKit/Platform/Logging.cpp:
```diff
@@ -27,6 +27,8 @@
 #include "config.h"
 #include "Logging.h"
 
+#include <wtf/NeverDestroyed.h>
```
Is this really needed?

pvollan replied:
This can be removed.

cdumez commented on Source/WebKit/Platform/WebKitLogClient.h:
```diff
@@ -0,0 +1,75 @@
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    WebKitLogClient(IPC::StreamClientConnection&, const LogStreamIdentifier&);
+    virtual ~WebKitLogClient() { }
```
Not needed.

cdumez commented on Source/WebKit/Platform/WebKitLogClient.h:
```diff
@@ -0,0 +1,75 @@
+    LogStreamIdentifier logStreamIdentifier() const { return m_logStreamIdentifier; }
+    Lock& logStreamLock() WTF_RETURNS_LOCK(m_logStreamLock) { return m_logStreamLock; }
+
+    bool isLogClientFinal() const final { return true; }
```
I think it can be private, it wouldn't make sense to call this on the subclass.

cdumez commented on Source/WebKit/Platform/WebKitLogClient.h:
```diff
@@ -0,0 +1,75 @@
+private:
+    Ref<IPC::StreamClientConnection> m_logStreamConnection;
+    LogStreamIdentifier m_logStreamIdentifier;
+    Lock m_logStreamLock;
```
It is still unclear to me what this lock is guarding against. Could you please clarify?

pvollan replied:
I believe the lock is required because `StreamClientConnection` is not thread safe. Before adding the lock, I was consistently getting debug asserts in `StreamServerConnection::didReceiveInvalidMessage` during layout tests. After adding the lock, the asserts have not appeared again.

cdumez commented on Source/WebCore/platform/WebCoreLogClient.h:
```diff
@@ -0,0 +1,45 @@
+    WebCoreLogClient() = default;
+    virtual ~WebCoreLogClient() { }
+
+    virtual bool isLogClientFinal() const { return false; }
```
This should be called `isWebKitLogClient()`

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -862,24 +862,34 @@ static void prewarmLogs()
 #if PLATFORM(IOS_FAMILY)
         prewarmLogs();
 #endif
+        logClient() = makeUnique<WebKitLogClient>(*logStreamConnection, logStreamIdentifier);
```
Should we ASSERT that `logClient()` is null before this?

cdumez commented on Source/WebKit/Shared/APIWebArchiveResource.h:
```diff
@@ -51,8 +51,8 @@ class WebArchiveResource : public API::ObjectImpl<API::Object::Type::WebArchiveR
     Ref<API::Data> data();
-    WTF::String URL();
-    WTF::String MIMEType();
+    WTF::String url();
```
Unclear to me why we're making these changes in this PR. They don't seem wrong but also not strictly related.

pvollan replied:
Ah, good catch! The PR is causing shifts in unified sources, which introduced some compile errors. After fixing the compile errors, I also fixed some style errors that were reported on the bots.

cdumez reviewed (approved):
LGTM

kkinnunen-apple commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -862,24 +862,35 @@ static void prewarmLogs()
+    auto* webkitLogClient = downcast<WebKitLogClient>(logClient().get());
+    if (!webkitLogClient)
+        return;
+    Locker locker { webkitLogClient->logStreamLock() };
```
I think here you could have:

```
webkitLogClient->log(channel, category, text, type)
```

And then have the implementation detail in the log client.
And it would make sense your base class support that interface too, so that you don't need  that kind of forced downcast.

pvollan replied:
Thanks for reviewing! I will address these issues shortly.

kkinnunen-apple commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -862,24 +862,35 @@ static void prewarmLogs()
         prewarmLogs();
 #endif
+        RELEASE_ASSERT(!logClient());
+        logClient() = makeUnique<WebKitLogClient>(*logStreamConnection, logStreamIdentifier);
```
I don't think anything guarantees that this reply comes back before first log..

kkinnunen-apple commented on Source/WebKit/Platform/WebKitLogClient.h:
```diff
@@ -0,0 +1,74 @@
+private:
+    bool isWebKitLogClient() const final { return true; }
+
+    Ref<IPC::StreamClientConnection> m_logStreamConnection;
```
This should be marked as `const Ref`. That way you can just use `m_connection`, instead of having those  accessors. Identifier, lock, connection should all likely be just implementation details of the client here.

The critical thing to guard is the connection. Thus you should mark that instance variable as `WTF_GUARDED_BY()`.. Everything else follows from that.

kkinnunen-apple commented on Source/WebCore/platform/WebCoreLogClient.h:
```diff
@@ -0,0 +1,45 @@
+    WebCoreLogClient() = default;
+    virtual ~WebCoreLogClient() { }
+
+    virtual bool isWebKitLogClient() const { return false; }
```
You probably should write the code so that no dynamic cast is needed, as suggested by below.

kkinnunen-apple commented on Source/WebCore/platform/WebCoreLogClient.h:
```diff
@@ -0,0 +1,45 @@
+#endif
+};
+
+WEBCORE_EXPORT std::unique_ptr<WebCoreLogClient>& logClient();
```
Personally I'd imagine this would be way simpler API to understand:
```
WEBCORE_EXPORT setLogClient(std::unique_ptr<WebCoreLogClient>&&);
WEBCORE_EXPORT extern WebCoreLogClient* logClient;

```

kkinnunen-apple commented on Source/WebCore/platform/WebCoreLogClient.h:
```diff
@@ -0,0 +1,45 @@
+
+namespace WebCore {
+
+class WebCoreLogClient {
```
Personally, WebCore::WebCoreLogClient seems a bit off. Also WebKit::WebKitLogClient : public WebCore::WebCoreLogClient.

kkinnunen-apple commented on Source/WebCore/platform/Logging.h:
```diff
@@ -28,6 +28,50 @@
+#include "WebCoreLogClient.h"
+
+#define RELEASE_LOG_FORWARDABLE(category, logMessage, ...) do { \
+    if (logClient()) \
```
These macros could cache the instance pointer.

kkinnunen-apple commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -862,24 +862,35 @@ static void prewarmLogs()
+
+    logStreamConnection->open(*this, RunLoop::protectedCurrent());
+
+    parentProcessConnection()->sendWithAsyncReply(Messages::WebProcessProxy::SetupLogStream(getpid(), WTFMove(serverHandle), logStreamIdentifier), [logStreamConnection, logStreamIdentifier] (IPC::Semaphore&& wakeUpSemaphore, IPC::Semaphore&& clientWaitSemaphore) {
```
I don't think WCP should be sending its pid to UI process. UI process knows it already.

kkinnunen-apple commented on Source/WebKit/Shared/LogStream.cpp:
```diff
@@ -43,12 +43,15 @@ namespace WebKit {
-        logStreamConnection->invalidate();
-    }
-    if (RefPtr logWorkQueue = m_logWorkQueue)
+    if (RefPtr logWorkQueue = m_logWorkQueue) {
```
This cannot be like this.
LogStream is message receiver, receiving messages in other thread.
It cannot be that the other thread is still using this, while this is also in destructor. Makes no sense.
This has to be its own function.
See the users  of ScopedActiveMessageReceiveQueue, you need to hold LogStream with it.

kkinnunen-apple commented on Source/WebKit/Shared/LogStream.cpp:
```diff
@@ -89,12 +92,17 @@ void LogStream::setup(uint64_t pid, IPC::StreamServerConnectionHandle&& serverCo
     m_logWorkQueue = IPC::StreamConnectionWorkQueue::create("Log work queue"_s);
     if (RefPtr logStreamConnection = m_logStreamConnection) {
-        logStreamConnection->open(*m_logWorkQueue);
+        if (RefPtr logWorkQueue = m_logWorkQueue)
```
All these if (workQueue) codes are incorrect, we cannot be in the code and not have the work queue, it makes no sense.

UI process cannot have a stream work queue per WCP. The queue should be a singleton.

pvollan committed:
Add logging definition file for OS log entries that should be forwarded from the WebContent process to the UI process
rdar://137601908
https://bugs.webkit.org/show_bug.cgi?id=282115

Reviewed by Chris Dumez.

Currently, we are forwarding preprocessed OS logs as strings from the WebContent process to the UI process in Lockdown Mode,
where they are sent to the logging daemon. Logging preprocessed strings is effectively disabling an important log size
optimization in the logging system. To preserve this optimization, all logging should use a literal format string, and
provide the data arguments when using the OS log API. This patch resolves this by adding logging definition files that are
used to generate code to send IPC messages from the WebContent process to the UI process, where the OS log API is called
with a literal format string, and the arguments provided over IPC. This patch adds one logging definition file for WebCore,
and one for WebKit, both with the name LogMessages.in.

This is the current format of the logging definition file:

<Identifier>, <Format string>, <Parameters>, <Log type (DEFAULT, INFO, ERROR, FAULT)>, <Category (Default, Process, Loading, etc)>

Here is an example:

DOCUMENTLOADER_STOPLOADING, "[pageID=%llu, frameID=%llu, isMainFrame=%d] DocumentLoader::stopLoading", (uint64_t, uint64_t, int), DEFAULT, Network

To add logging in the WebContent process that will be forwarded to the UI process in the efficient way, you will need
to add an entry to one of the logging definition files, and then pass the identifier to one of the new logging macros
- RELEASE_LOG_FORWARDABLE, RELEASE_LOG_ERROR_FORWARDABLE, etc.

Based on the logging definition files, the file LogStream.messages.in is generated. This file is used as input to the
standard script that generates code to send messages over IPC. The log forwarding is using streaming IPC, which is
based on shared memory, and is very efficient.

Additionally, two new log client classes are created, WebCore::LogClient and WebKit::LogClient. The scripts added in this
patch will generate virtual functions in these classes for every log message in the definitions file. Using one of the
new logging macros (RELEASE_LOG_FORWARDABLE, etc.), will result in calling one of these virtual functions, which will
send the IPC to the UI process. WebKit::LogClient is inheriting from WebCore::LogClient, and is implementing the sending
of IPC for both WebCore and WebKit logs.

This patch enables log forwarding for all Cocoa platforms, not just in Lockdown Mode.

Page load benchmarks results are neutral with this change for the configurations tested so far, and the sample count
is also neutral in PLT, which indicates that there should be no power regression from this change. Power benchmark
results are neutral on both iOS and macOS.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::m_remote):
(WebCore::HTMLMediaElement::~HTMLMediaElement):
(WebCore::HTMLMediaElement::insertedIntoAncestor):
(WebCore::HTMLMediaElement::didFinishInsertingNode):
(WebCore::HTMLMediaElement::removedFromAncestor):
(WebCore::HTMLMediaElement::currentMediaTime const):
(WebCore::HTMLMediaElement::mediaPlayerTimeChanged):
(WebCore::HTMLMediaElement::mediaPlayerRateChanged):
(WebCore::HTMLMediaElement::mediaPlayerEngineUpdated):
(WebCore::HTMLMediaElement::updatePlayState):
(WebCore::HTMLMediaElement::visibilityStateChanged):
(WebCore::HTMLMediaElement::setShouldDelayLoadEvent):
(WebCore::HTMLMediaElement::setBufferingPolicy):
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::mediaPlayerRenderingModeChanged):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::stopLoading):
(WebCore::DocumentLoader::attachToFrame):
(WebCore::DocumentLoader::detachFromFrame):
(WebCore::DocumentLoader::startLoadingMainResource):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::changeLocation):
(WebCore::FrameLoader::loadURLIntoChildFrame):
(WebCore::FrameLoader::prepareForLoadStart):
(WebCore::FrameLoader::loadFrameRequest):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::loadWithNavigationAction):
(WebCore::FrameLoader::loadWithDocumentLoader):
(WebCore::FrameLoader::stopAllLoaders):
(WebCore::FrameLoader::setDocumentLoader):
(WebCore::FrameLoader::setPolicyDocumentLoader):
(WebCore::FrameLoader::setProvisionalDocumentLoader):
(WebCore::FrameLoader::setState):
(WebCore::FrameLoader::transitionToCommitted):
(WebCore::FrameLoader::checkLoadCompleteForThisFrame):
(WebCore::FrameLoader::continueLoadAfterNavigationPolicy):
* Source/WebCore/loader/ProgressTracker.cpp:
(WebCore::ProgressTracker::progressStarted):
(WebCore::ProgressTracker::progressCompleted):
(WebCore::ProgressTracker::finalProgressComplete):
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::willSendRequestInternal):
* Source/WebCore/loader/SubresourceLoader.cpp:
(WebCore::SubresourceLoader::willSendRequestInternal):
(WebCore::SubresourceLoader::didReceiveResponse):
(WebCore::SubresourceLoader::didFinishLoading):
(WebCore::SubresourceLoader::didFail):
(WebCore::SubresourceLoader::willCancel):
* Source/WebCore/platform/Logging.cpp:
(WebCore::logClient):
* Source/WebCore/platform/Logging.h:
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::addSession):
(WebCore::PlatformMediaSessionManager::removeSession):
(WebCore::PlatformMediaSessionManager::sessionWillEndPlayback):
(WebCore::PlatformMediaSessionManager::sessionCanProduceAudioChanged):
(WebCore::PlatformMediaSessionManager::maybeActivateAudioSession):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::updateSessionState):
(WebCore::MediaSessionManagerCocoa::clientCharacteristicsChanged):
(WebCore::MediaSessionManagerCocoa::sessionCanProduceAudioChanged):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::didReceiveInvalidMessage):
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Platform/Sources.txt:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/Scripts/generate-message-receiver.py:
(main):
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/API/c/mac/WKWebArchiveResource.cpp:
(WKWebArchiveResourceCopyURL):
(WKWebArchiveResourceCopyMIMEType):
* Source/WebKit/Shared/APIWebArchiveResource.h:
* Source/WebKit/Shared/APIWebArchiveResource.mm:
(API::WebArchiveResource::create):
(API::WebArchiveResource::WebArchiveResource):
(API::WebArchiveResource::url):
(API::WebArchiveResource::mimeType):
(API::WebArchiveResource::textEncoding):
(API::WebArchiveResource::URL): Deleted.
(API::WebArchiveResource::MIMEType): Deleted.
* Source/WebKit/Shared/LogStream.cpp:
(WebKit::LogStream::~LogStream):
(WebKit::LogStream::setup):
* Source/WebKit/Shared/LogStream.h:
(WebKit::LogStream::create):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::setupLogStream):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::finishLaunchingProcess):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::scheduleLoad):
(WebKit::WebLoaderStrategy::scheduleLoadFromNetworkProcess):
* Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
(WebKit::WebResourceLoader::WebResourceLoader):
(WebKit::WebResourceLoader::willSendRequest):
(WebKit::WebResourceLoader::didReceiveResponse):
(WebKit::WebResourceLoader::didReceiveData):
(WebKit::WebResourceLoader::didFinishResourceLoad):
(WebKit::WebResourceLoader::serviceWorkerDidNotHandle):
(WebKit::WebResourceLoader::didFailResourceLoad):
(WebKit::WebResourceLoader::didBlockAuthenticationChallenge):
(WebKit::WebResourceLoader::stopLoadingAfterXFrameOptionsOrContentSecurityPolicyDenied):
(WebKit::WebResourceLoader::didReceiveResource):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::navigationActionData const):
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/cocoa/LaunchServicesDatabaseManager.mm:
(WebKit::LaunchServicesDatabaseManager::handleEvent):
(WebKit::LaunchServicesDatabaseManager::waitForDatabaseUpdate):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformInitializeWebProcess):
(WebKit::WebProcess::setupLogStream):
(WebKit::WebProcess::sendLogOnStream):
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

Canonical link: https://commits.webkit.org/287665@main
